### PR TITLE
Migration clean up and fixes for 4.12 (close #88)

### LINKED ISSUE #88: 4.12 Client Side Testing - v4 Migration Cleanup
- [ ] "wikilinks" during migration the OWASP Mediawiki links were tagged "wikilink" these needs to be converted to proper absolute URLs.
  - Unless the wikilink is to content within the OTG, in which case it should be the proper new github relative URL (URL Encoded).
- [ ] Foot notes/references should use standard in-line links. (One of the first two here: https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#links)
    - If a foot note is really required use numeric reference type tags, ex: Content ref: `[[1]]` Footer ref:  `[1]: http://example.org`
- [ ] Images need to be saved to <chapter (or sub-section)>/images/<image.ext> and linked relatively like images/image.ext. (Images in a folder along side the relevant content. [Even if that means duplicates in the overall project.])
- [ ] Any spurious `/` terminating lines, etc. need to be removed and appropriately fixed up for markdown (it's likely a spot that requires a double line-feed prior to an image or other element).
- [ ] In some cases (due to the source of the text) there are broken apostrophes `'` they should just be restored.
- [ ] HTML entities need to be properly handled.
  - Occurrences of `\$` should become `&dollar;`
  - At symbols `@` should become `&commat;`
- [ ] Headers should be properly formatted with leading # (hashtags) per: https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers

### LINKED ISSUE #160: 4.12.9 Edit "How to Test" section title
Change the "How to Test" section title from H3 to H2 to improve standardization across all section titles (the other section titles are all H2).

### PR DESCRIPTION
**Pull request by victoriadrake**
This PR covers issue #88 

- [x] This PR handles the issue and requires no additional PRs.
- [x] You have validated the need for this change.

This PR:

- Fixes wiki artifacts and transitions to Markdown.
- Fixes or implements code formatting where appropriate.
- Removes dead or 404 links.
- Removes links with no direct reference in the chapter text.
- Removes outdated screenshots in favor of links to instructions (4.12.12)
- Corrects copy and removes of out-of-scope sections, i.e. repeated references to black-box testing not being applicable.
- Aligns the touched sections with the article template.
- Makes filenames consistent.
- Change 4.12.12 "Test Local Storage" to "Testing Web Storage" to alleviate confusion between the title and `localStorage`. Also mostly rewrite section. Sorry.
- Removes non-existent quote from 4.12.13 ([10]).
- Closes #160.

This PR does not address these concerns:

- Link for domxss.com have no https and are a bit old
- Some links to code.google.com documents go to pages with broken Markdown
- The value and accuracy of the content may be outdated
- Some sections do not appear to have actual testing instructions:
  - 4.12.2 Testing for JavaScript Execution
  - 4.12.3 Testing for HTML Injection
- The section 4.12.9 Testing for Clickjacking should not cover defenses, and currently duplicates material from the [OWASP Clickjacking Defense Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Clickjacking_Defense_Cheat_Sheet.html) - thus the lines referenced in #183 have not been changed
- Not sure that the JS examples in 4.12.13 are syntactically correct

I generally sought to get this section to a place where contributors can begin to add to it and update the content that is still relevant, instead of still cleaning it up.

### CODE REVIEW

**Comment by kingthorin** [comment]
I'll do a more thorough review of this in the coming days, however after a quick skim: Please make sure all headings use title caps.

**Comment by kingthorin** [comment]
Thanks!

File: document/4_Web_Application_Security_Testing/4.12_Client_Side_Testing/4.12.7_Testing_Cross_Origin_Resource_Sharing_OTG-CLIENT-007.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+### Other Headers
+
+There are other headers involved like Access-Control-Max-Age that determines the time a preflight request can be cached in the browser, or Access-Control-Expose-Headers that indicates which headers are safe to expose to the API of a CORS API specification, both are response headers specified in the CORS W3C document.
+
+## How to Test
+
+A tool such as [OWASP Zed Attack Proxy Project](https://www.owasp.org/index.php/OWASP_Zed_Attack_Proxy_Project) can enable testers to intercept HTTP headers, which can reveal how CORS is used. Testers should pay particular attention to the origin header to learn which domains are allowed. Also, manual inspection of the JavaScript is needed to determine whether the code is vulnerable to code injection due to improper handling of user supplied input. Below are some examples:
+
+### Example 1: Insecure Response with Wildcard '\*' in Access-Control-Allow-Origin
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.12_Client_Side_Testing/4.12.7_Testing_Cross_Origin_Resource_Sharing_OTG-CLIENT-007.md]: remove the `\` before the `*`

File: document/4_Web_Application_Security_Testing/4.12_Client_Side_Testing/4.12.7_Testing_Cross_Origin_Resource_Sharing_OTG-CLIENT-007.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+Access-Control-Allow-Origin: *
+Vary: Accept-Encoding
+Content-Length: 92
+Keep-Alive: timeout=15, max=100
+Connection: Keep-Alive
+Content-Type: text/html
+
+Injected Content from attacker.bar <img src="#" onerror="alert('Domain: '+document.domain)">
+```
+j
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.12_Client_Side_Testing/4.12.7_Testing_Cross_Origin_Resource_Sharing_OTG-CLIENT-007.md]: Rogue character :)

File: document/4_Web_Application_Security_Testing/4.12_Client_Side_Testing/4.12.9_Testing_for_Clickjacking_OTG-CLIENT-009.md [hunk 3]
```diff
@@ -1,50 +1,47 @@
...
+## Summary
 
-A Clickjacking attack uses seemingly innocuous features of HTML and Javascript to force the victim to perform undesired actions, such as clicking a button that appears to perform another operation. This is a “client side” security issue that affects a variety of browsers and platforms.
+“Clickjacking,” a subset of “UI redressing,” is a malicious technique whereby a web user is deceived into interacting (in most cases by clicking) with something other than what the user believes they are interacting with. This type of attack, either alone or in conjunction with other attacks, could potentially send unauthorized commands or reveal confidential information while the victim is interacting with seemingly-harmless web pages. The term “clickjacking” was coined by Jeremiah Grossman and Robert Hansen in 2008.
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.12_Client_Side_Testing/4.12.9_Testing_for_Clickjacking_OTG-CLIENT-009.md]: no need for quotes anywhere in here.

File: document/4_Web_Application_Security_Testing/4.12_Client_Side_Testing/4.12.9_Testing_for_Clickjacking_OTG-CLIENT-009.md [hunk 4]
```diff
@@ -1,50 +1,47 @@
...
 
-To carry out this type of technique the attacker has to create a seemingly harmless web page that loads the target application through the use of an iframe (suitably concealed with CSS code). Once this is done, the attacker could induce the victim to interact with his fictitious web page by other means (through, for example, social engineering). Like other attacks, a common prerequisite is that the victim is authenticated against the attacker’s target website.
+A clickjacking attack uses seemingly-innocuous features of HTML and JavaScript to force the victim to perform undesired actions, such as clicking an invisible button that performs an unintended operation. This is a client side security issue that affects a variety of browsers and platforms.
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.12_Client_Side_Testing/4.12.9_Testing_for_Clickjacking_OTG-CLIENT-009.md]: easier on the english, no need to use `innocuous`. Pretty rare word to be found anywhere, and non native speakers will find it hard.

File: document/4_Web_Application_Security_Testing/4.12_Client_Side_Testing/4.12.1_Testing_for_DOM-based_Cross_Site_Scripting_OTG-CLIENT-001.md [hunk 5]
```diff
@@ -1,105 +1,83 @@
...
 Due to their nature, DOM-based XSS vulnerabilities can be executed in many instances without the server being able to determine what is actually being executed. This may make many of the general XSS filtering and detection techniques impotent to such attacks.
 
-The first hypothetical example uses the following client side code:
+This hypothetical example uses the following client side code:
 
+```html
 <script>
-`document.write("Site is at: " + document.location.href + `“`.`”`);`\
-
+`document.write("Site is at: " + document.location.href + `“`.`”`);`
```
[the remark this answers is not in the document]
Comment (victoriadrake) [reply document/4_Web_Application_Security_Testing/4.12_Client_Side_Testing/4.12.1_Testing_for_DOM-based_Cross_Site_Scripting_OTG-CLIENT-001.md]: Doh. Nice catch.

File: document/4_Web_Application_Security_Testing/4.12_Client_Side_Testing/4.12.7_Testing_Cross_Origin_Resource_Sharing_OTG-CLIENT-007.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+Access-Control-Allow-Origin is a response header used by a server to indicate which domains are allowed to read the response. Based on the CORS W3 Specification it is up to the client to determine and enforce the restriction of whether the client has access to the response data based on this header.
+
+From a penetration testing perspective you should look for insecure configurations as for example using a '\*' wildcard as value of the Access-Control-Allow-Origin header that means all domains are allowed. Other insecure example is when the server returns back the origin header without any additional checks, what can lead to access of sensitive data. Note that this configuration is very insecure, and is not acceptable in general terms, except in the case of a public API that is intended to be accessible by everyone.
```
[the remark this answers is not in the document]
Comment (victoriadrake) [reply document/4_Web_Application_Security_Testing/4.12_Client_Side_Testing/4.12.7_Testing_Cross_Origin_Resource_Sharing_OTG-CLIENT-007.md]: Those should also be backticks.

File: document/4_Web_Application_Security_Testing/4.12_Client_Side_Testing/4.12.7_Testing_Cross_Origin_Resource_Sharing_OTG-CLIENT-007.md (hunk 2 above)
Comment (victoriadrake) [reply document/4_Web_Application_Security_Testing/4.12_Client_Side_Testing/4.12.7_Testing_Cross_Origin_Resource_Sharing_OTG-CLIENT-007.md]: Mister J!

File: document/4_Web_Application_Security_Testing/4.12_Client_Side_Testing/4.12.10_Testing_WebSockets_OTG-CLIENT-010.md [hunk 7]
```diff
@@ -1,92 +1,48 @@
...
-   * WebSockets do not handle authorization, normal black-box authorization tests should be carried out. Refer to the [Authorization Testing](https://www.owasp.org/index.php/Testing_for_Authorization) sections of this guide.
+### 3. Examine Confidentiality and Integrity
 
-6. Input Sanitization.
-   * Use [OWASP Zed Attack Proxy](https://www.owasp.org/index.php/OWASP_Zed_Attack_Proxy_Project) (ZAP)'s WebSocket tab to replay and fuzz WebSocket request and responses. Refer to the [Testing for Data Validation](https://www.owasp.org/index.php/Testing_for_Data_Validation) sections of this guide.
+- Check that the WebSocket connection is using SSL to transport sensitive information via the `wss://` URI scheme.
+- Check the SSL implementation for security issues (Valid Certificate, BEAST, CRIME, RC4, etc). Refer to the [Testing for Weak SSL/TLS Ciphers, Insufficient Transport Layer Protection (OTG-CRYPST-001)](https://www.owasp.org/index.php/Testing_for_Weak_SSL/TLS_Ciphers,_Insufficient_Transport_Layer_Protection_(OTG-CRYPST-001)) section of this guide.
 
-#### Example 1
+### 4. Use a Testing Tool
```
[the remark this answers is not in the document]
Comment (victoriadrake) [reply document/4_Web_Application_Security_Testing/4.12_Client_Side_Testing/4.12.10_Testing_WebSockets_OTG-CLIENT-010.md]: I'm afraid I don't know what you mean. What tests, and what "other guide"?

Can you mock up an example?

[the remark this answers is not in the document]
Comment (ThunderSon) [reply document/4_Web_Application_Security_Testing/4.12_Client_Side_Testing/4.12.10_Testing_WebSockets_OTG-CLIENT-010.md]: You scrapped out from the old document, this one, but the older version, points 4, 5, and 6 (AuthN, AuthZ, Sanitization). You replaced them with `Use a Testing Tool`. It actually removed content and didn't replace with anything beneficial. Any tester will be definitely using a tool. But what about it?

[the remark this answers is not in the document]
Comment (victoriadrake) [reply document/4_Web_Application_Security_Testing/4.12_Client_Side_Testing/4.12.10_Testing_WebSockets_OTG-CLIENT-010.md]: Oh, I see. I removed 4 and 5 as they said “WebSockets do not handle authentication/authorization” so it didn’t seem to be relevant to the testing instructions. I kept 6 in essence, only changed the title to describe its content (using ZAP). It was previously titled “Input Sanitization” and I thought that, as a heading, it didn’t seem to describe the instructions that followed.

I can replace the links for Authentication, Authorization, and Data Validation, if you think these are good places for them; or I can rewind my changes so they’re not holding back this PR, and we can revisit them in another issue.

[the remark this answers is not in the document]
Comment (ThunderSon) [reply document/4_Web_Application_Security_Testing/4.12_Client_Side_Testing/4.12.10_Testing_WebSockets_OTG-CLIENT-010.md]: Yes please, let's not remove content that we didn't improve, even if it's not well written or not properly taken care of.

**Comment by ThunderSon** [comment]
@victoriadrake I believe you missed 3 comments. Once you go over them again please let me know.

**Comment by victoriadrake** [comment]
@ThunderSon I've rewound the requested changes and fixed conflicts. Please have another look.

**Comment by ThunderSon** [comment]
You've done a great job! My last comment would be about using `*` versus `-` for unordered lists. @rejahrehim we last agreed on `-`, right? If so, @victoriadrake please let's go back to having `-` for those lists, since it will be added as a linter rule later on.

**Comment by victoriadrake** [comment]
@ThunderSon There ya go!

**Pull request merged**